proposals/pattern-const-expressions.md

```diff
@@ -0,0 +1,120 @@
+# Constant `is` pattern expressions
+
+* [x] Proposed
+* [ ] Prototype: None
+* [ ] Implementation: None
+* [ ] Specification: Started, below
+
+## Summary
+[summary]: #summary
+
+Consider `is` pattern expressions as constant if the LHS is constant.
+
+## Motivation
+[motivation]: #motivation
+
+Certain pattern expressions that can be converted into simpler boolean expressions using equality and comparison operators are never considered constant expressions, even when all parts of the expression are considered constant and can be evaluated during compilation. However, the lowered versions of those expressions (that only involve equality and comparison operators) are always considered constant. This prohibits the ability to utilize `is` expressions for operations like comparing against a range (e.g. `x is >= 'a' and 'z'`), or checking against distinct values (e.g. `x is Values.A or Values.B or Values.C`).
+
+## Detailed design
+[design]: #detailed-design
+
+### Spec
+[spec]: #spec
+
+The [section](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1223-constant-expressions) in the spec needs to be updated accordingly, in the following segments:
+
+> Only the following constructs are permitted in constant expressions:
+> - Literals (including the null literal).
+> - References to const members of class and struct types.
+> - References to members of enumeration types.
+> - References to local constants.
+> - Parenthesized subexpressions, which are themselves constant expressions.
+> - Cast expressions.
+> - checked and unchecked expressions.
+> - nameof expressions.
+> - The predefined +, –, !, and ~ unary operators.
+> - The predefined +, –, *, /, %, <<, >>, &, |, ^, &&, ||, ==, !=, <, >, <=, and >= binary operators.
+> - The ?: conditional operator.
+> - **Pattern expressions with only the following subpatterns:**
+>   - **Boolean literal patterns.**
+>   - **Numeric literal patterns.**
+>   - **Character literal patterns.**
+>   - **String literal patterns.**
+>   - **Relative numeric literal patterns.**
+>   - **Relative character literal patterns.**
+>   - **Null literal patterns.**
+>   - **Default literal patterns.**
+> - sizeof expressions, provided the unmanaged-type is one of the types specified in §23.6.9 for which sizeof returns a constant value.
+> - Default value expressions, provided the type is one of the types listed above.
+
+The allowed subpatterns as shown in the list above are called "constant subpatterns", as they will be eligible for compile-time evaluation.
```

Comment:
presumably the `not`, `and` and `or` patterns should also work.

Reply:
i'd enumerate the exact set of patterns we expect this to work for (same as how we enumerate the exact set of unary and binary operators, etc.).  note: based on how the constant-expressions spec works out, i think everything else will likely fall out.

You can then give examples for each case.  I don't htink you have to specify how it evaluates as the constant expression section already says this: 

> Whenever an expression fulfills the requirements listed above, the expression is evaluated at compile-time. This is true even if the expression is a subexpression of a larger expression that contains non-constant constructs.

> The compile-time evaluation of constant expressions uses the same rules as run-time evaluation of non-constant expressions, except that where run-time evaluation would have thrown an exception, compile-time evaluation causes a compile-time error to occur.

--

Note: @333fred if we're doing this spec, would it makes sense to roll in `switch` expressions at the same time, given that the lang already supports `?:`?  Or would you prefer to keep that excluded?  Thanks!

Reply:
LDM did express interest in it, but didn't directly approve it. Probably best to leave it as an open question for now.

Reply:
That works for me.  We could bring for an LDM read and easily add if desired.

Reply:
Note, there is a discussion for the equivalent extension to support switch expressions in #7489, which will also be noted in the spec doc